File: docs/clause-and-effect/in-prolog/worksheet-02-drinking-pairs.pl
drinks(john, martini).
drinks(mary, gin).
drinks(susan, vodka).
drinks(john, gin).
drinks(fred, gin).

%% pair(X, Y, Z) :-
%%   drinks(X, Z), drinks(Y, Z).

pair(X, Y, Z) :-
  drinks(X, Z), drinks(Y, Z), X \== Y.

%% pair(X, john, martini).
%% pair(mary, susan, gin).
%% pair(john, mary, gin).
%% pair(john, john, gin).
%% pair(X, Y, gin).
%% pair(bertram, lucinda, vodka).
%% pair(X, Y, Z).

File: docs/old/clause-and-effect/worksheet-02-drinking-pairs.js
let logic = require ("logic-db")

let drinks = new logic.db_t

drinks.i ({ person: "john",  alcohol: "martini" })
drinks.i ({ person: "mary",  alcohol: "gin" })
drinks.i ({ person: "susan", alcohol: "vodka" })
drinks.i ({ person: "john",  alcohol: "gin" })
drinks.i ({ person: "fred",  alcohol: "gin" })
drinks.i ({ person: "fred",  alcohol: "vodka" })

let pair = new logic.db_t

pair.i ({
    p1: "?p1",
    p2: "?p2",
    alcohol: "?alcohol",
}) .cond ((the) => {
    return drinks.o ({ person: the.p1, alcohol: the.alcohol })
        .and (drinks.o ({ person: the.p2, alcohol: the.alcohol }))
        .not_eqv (the.p1, the.p2)
})

pair.query_log (100) ({
    p1: "?x",
    p2: "mary",
    alcohol: "gin",
})

pair.query_log (100) ({
    p1: "?x",
    p2: "?y",
    alcohol: "gin",
})

pair.query_log (100) ({
    p1: "?x",
    p2: "?y",
    alcohol: "?alcohol",
})
